README.md
# INDELfindR

## About
INDELfindR is an R based command line tool for detecting simple and complex insertion deletion (INDEL) variants which outputs variant calls in a VCF v4.3 file compatible with downstream analysis and annotation tools.

## Setup

#### Requirements

INDELfindR is compatible with Linux, MacOS, and Windows. 

*Note:* INDELfindR does not support parallel processing on Windows.

Depends: R (≥ 4.1.0) 

INDELfindR has the following R package dependencies:

The following packages can be installed from CRAN:

require(devtools)
install_version("<each_package>", version = "<version_number")

```
argparse/2.1.3
plyr/1.8.6
tidyverse/1.3.1
inline/0.3.19
bettermc/1.1.2
purrr/0.3.4
dplyr/1.0.8
```

While these packages can be installed manually via the Bioconductor 3.14 release:

First, install Bioconductor v3.14

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(version = "3.14")

Then install the following packages with Bioconductor 3.14:

BiocManager::install(c("GenomicFeatures", "BSgenome.Hsapiens.UCSC.hg38","bamsignals",GenomicAlignments))

The package versions which are installed via Bioconductor 3.14 should be as follows:
```
GenomicFeatures/1.46.1
BSgenome.Hsapiens.UCSC.hg38/1.4.4
bamsignals/1.26.0
GenomicAlignments/1.30.0
```
#### Installation

Install the INDELfindR R package from CRAN to install INDELfindR and all of it's dependencies.

(This will work once INDELfindR package is registered. Until then they must be downloaded manually via R.)

```
install.packages("INDELfindR")

# test installation
library(INDELfindR)
```

#### Run INDELfindR with Demo Data
We provide a test bam file for running with your INDELfindR setup. We also provide two output files containing the expected output for you to compare your test run results against.

To run INDELfindR with the demo data we provide, download both the bam file, named EGFR_mutations_indelfindr_demo.bam and the .bai index file, named EGFR_mutations_indelfindr_demo.bam.bai, from the /demo/ directory of our INDELfindR Github repository. Make sure the .bai index file is downloaded after downloading the bam file since the program performs an internal check to make sure that the index file creation date is after the creation data of the bam file. 

Once you've installed the indelfindr R package following the above instructions, you can download the INDELfindR executable Rscript stored in the /INDELfindR_script/ directory of the INDELfindR Github repository, and run it on the command line. To run the demo, run the following command within the directory containing the indelfindr.R Rscript:

```
Rscript indelfindr -a /<your_path_to_file>/EGFR_mutations_indelfindr_demo.bam -t /<your_path_to_file>/EGFR_regions_of_interest.txt -o EGFR_mutations_indelfindr_demo

```

The output files for this analysis that we provide for your comparison are stored inside the same /demo/ directory and are named EGFR_mutations_indelfindr_demo.vcf and
EGFR_mutations_indelfindr_demo.indel.cigars.csv.

If the results you recieve are the same as those in our provided demo results files, you are ready to run your analysis!

## Quickstart

INDELfindR accepts indexed BAM files as input. INDELfindR only accepts human DNA sequence data which has been aligned to the hg38 reference genome version.

After installing the INDELfindR R package, download the indelfindr.R script found [here](https://github.com/TranslationalBioinformaticsLab/INDELfindR/blob/main/src/INDELfindR.R) and run INDELfindR from the command line by calling:

```
Rscript indelfindr.R -a <indexed_bam.bam> (options)
```

## Usage Instructions

```
usage: indelfindr.R [-h] [-v] [-q] -a filename [-f number] [-b number]
                    [-l number] [-nr number] [-mq number] [-t filename]
                    [-nc number] [-p] [-vaf number] [-dp number] [-o filepath]

optional arguments:
  -h, --help            show this help message and exit
  -v, --verbose         Print extra output [default]
  -q, --quietly         Print little output
  -a filename, --alignment_file filename
                        Filepath to alignment file in bam format
  -f number, --flanking_region_length number
                        Minimum number of `=` or `X` CIGAR operators required
                        to flank a string of nearby `I` `D` operators in order
                        for a call to be made
  -b number, --bam_bin_size number
                        Length of non-overlapping sliding windows to use to
                        extract overlapping reads from bam file and store in
                        memory at a time (default = 100000). Larger windows
                        require loading more reads into memory at one time,
                        while smaller windows require longer runtime due to
                        evaluating more reads which overlap neighboring
                        sliding windows and removing more duplicate read calls
  -l number, --min_indel_length number
                        Minumum number of ref or alt allele basepairs for an
                        indel to be called (default = 5 nucleotides)
  -nr number, --number_reads number
                        Minumum number of supporting reads for a single given
                        indel to be called
  -mq number, --mapq_filter number
                        Minumum read mapping quality required to evaluate a
                        read.
  -t filename, --target_regions filename
                        File path to .bed file containing regions in which to
                        perform variant calling. Chromosome syntax must match
                        bed file (ex. `Chr1`, `chr1`, or `1`).
  -nc number, --number_cores number
                        Number of cores to use for fork/join parallel
                        computation during indel calling and filtering.
  -p, --primary_chromosomes
                        Call indels in primary chromosomes only ignoring ALT
                        contig alignments (chr1-22,X,Y,M only).
  -vaf number, --vaf_filter number
                        Minumum variant allele freqency required for an indel
                        to be reported
  -dp number, --read_depth_filter number
                        Minumum indel range read depth required for an indel
                        to be reported
  -o filepath, --outname filepath
                        Define the output directory path
```

## Running INDELfindR with ANNOVAR Variant Annotation

We provide a shell script template that you may use to call indels with INDELfindR and annotate the resulting VCF output file with ANNOVAR databases on Unix systems. You may download this shell script template [here]().

## ANNOVAR Setup for INDELfindR

For users' convenience we've outlined a quickstart tutorial to set up the ANNOVAR databases used in our provided INDELfindR with ANNOVAR shell script.

### Set up COSMIC, ClinVar, and dpsnp databases:

Install ANNOVAR and prepare ANNOVAR databases using this tutorial:
https://annovar.openbioinformatics.org/en/latest/user-guide/startup/

### Set Up Some Suggested Annovar Databases

Note: avsnp150 (below) and is a left aligned dbsnp database for use with left aligned vcf. 

I chose avsnp150 from ANNOVAR, instead of downloading dbSNP build 155 and preparing a database from scratch, since it includes annovar indexes for faster annotation.

```bash

annotate_variation.pl -buildver hg38 -downdb -webfrom annovar avsnp150 humandb/

annotate_variation.pl -buildver hg38 -downdb -webfrom annovar clinvar_20210501 humandb/

annotate_variation.pl -buildver hg38 -downdb -webfrom annovar refGene humandb/

```

### Custom Database Build: Current COSMIC v94 database

Since COSMIC70 is the latest version available through Annovar directly, set up v94 database.

See this ink for complete instructions: https://annovar.openbioinformatics.org/en/latest/user-guide/filter/#cosmic-annotations

I've included my commands below but I would use the link above for complete instructions.

#### Get Download Authorization Code
echo "youremail@brown.edu:your_email_password" | base64

(this produces a base64 code which you can copy and paste into the commands below)

#### Download coding variants

curl -H "Authorization: Basic Z2VvcmdlX3RvbGxlZnNvbkBicm93bi5lZHU6TmV3cG9ydDk0IQo=" https://cancer.sanger.ac.uk/cosmic/file_download/GRCh38/cosmic/v94/VCF/CosmicCodingMuts.vcf.gz

curl "https://cog.sanger.ac.uk/cosmic/GRCh38/cosmic/v94/VCF/CosmicCodingMuts.vcf.gz?AWSAccessKeyId=KRV7P7QR9DL41J9EWGA2&Expires=1633980713&Signature=2PL9NJwDJpqbZflB2zeo1V%2BNDrU%3D" > CosmicCodingMuts.vcf

#### Download non-coding variants
curl -H "Authorization: Basic Z2VvcmdlX3RvbGxlZnNvbkBicm93bi5lZHU6TmV3cG9ydDk0IQo=" https://cancer.sanger.ac.uk/cosmic/file_download/GRCh38/cosmic/v94/VCF/CosmicNonCodingVariants.vcf.gz

curl "https://cog.sanger.ac.uk/cosmic/GRCh38/cosmic/v94/VCF/CosmicNonCodingVariants.vcf.gz?AWSAccessKeyId=KRV7P7QR9DL41J9EWGA2&Expires=1633981220&Signature=X8JpUPgshA3PtIesTXZejadaNEw%3D" > CosmicNonCodingMuts.vcf.gz

#### Download mutant export
curl -H "Authorization: Basic Z2VvcmdlX3RvbGxlZnNvbkBicm93bi5lZHU6TmV3cG9ydDk0IQo=" https://cancer.sanger.ac.uk/cosmic/file_download/GRCh38/cosmic/v94/CosmicMutantExport.tsv.gz

curl "https://cog.sanger.ac.uk/cosmic/GRCh38/cosmic/v94/CosmicMutantExport.tsv.gz?AWSAccessKeyId=KRV7P7QR9DL41J9EWGA2&Expires=1633981344&Signature=RtjmsnVfZreiHz7f5nPA%2BsW13%2Fo%3D" > CosmicMutantExport.tsv.gz

#### Download noncoding mutant export

curl -H "Authorization: Basic Z2VvcmdlX3RvbGxlZnNvbkBicm93bi5lZHU6TmV3cG9ydDk0IQo=" https://cancer.sanger.ac.uk/cosmic/file_download/GRCh38/cosmic/v94/CosmicNCV.tsv.gz

curl "https://cog.sanger.ac.uk/cosmic/GRCh38/cosmic/v94/CosmicNCV.tsv.gz?AWSAccessKeyId=KRV7P7QR9DL41J9EWGA2&Expires=1634050162&Signature=bfa1NUzyHH1uzprzRDiE1AUKthk%3D" > CosmicNCV.tsv.gz

#### Prepare COSMIC database for ANNOVAR

```bash

perl ./prepare_annovar_user.pl -dbtype cosmic /path_to_your_directory/cosmic_database/CosmicMutantExport.tsv -vcf /path_to_your_directory/cosmic_database/CosmicCodingMuts.vcf.gz > ./hg38_cosmic94.txt

```

###### Coding
```
perl ./prepare_annovar_user.pl -dbtype cosmic CosmicMutantExport.tsv -vcf CosmicCodingMuts.vcf > hg38_cosmic94_coding.txt 
```
##### Noncoding
```
perl ./prepare_annovar_user.pl -dbtype cosmic CosmicNCV.tsv -vcf CosmicNonCodingMuts.vcf > hg38_cosmic94_noncoding.txt
```


## Running INDELfindR with Docker

### Installing Docker
You can run INDELfindR using the Docker images without installing INDELfindR, R, and it's dependencies. 

To run INDELfindR from a Docker image, first [install Docker](https://docs.docker.com/install/).

Once Docker is installed, double-click the Docker.app in the Applications folder to start Docker Desktop. A whale icon will appear in the top status bar to indicate that Docker is running and accessible from the terminal. You can quit Docker once you are finished using INDELfindR by clicking the Docker whale icon in the top status bar and clicking "Quit Docker Desktop."

### Running INDELfindR with Docker

Once Docker is running, you can run INDELfindR by running the Docker commands below in the Mac/Linux terminal or Windows PowerShell. All INDELfindR command line arguments may be specified after the  call to `indelfindr` in the following `docker run` commands.

Example run on Mac or Linux:

```
docker run --cpuset-cpus="0-1" -v "$PWD":/data indelfindr -a /data/test.sorted.bam -o test.sorted -b 100000000 (options)
```

Example run on Windows:
*Reminder*: --number_cores is set to 1 . (In `docker run`, `--cpuset-cpus="0"` specifies  the first core - core "0")
```
docker run --cpuset-cpus="0" -v ${pwd}:/data indelfindr -a /data/test.sorted.bam -o test.sorted -nc 1 -b 100000000 (options)
```

NEWS.md
## version 1.0.0

---


## version 0.0.0.9000

---

### NEWS.md setup

- added NEWS.md creation with newsmd



src/INDELfindR.R
#!/usr/bin/env Rscript

#############################################################################################
#
# Script to call indels using refined cigar strings in non-overlapping sliding windows
# genome wide or across the user specified genomic regions and output indel calls in VCF 4.3.
#
#############################################################################################

logo <- r"{ 
             _   _    ____  U _____ u  _       _____              _   _    ____    ____     
    ___     | \ |"|  |  _"\ \| ___"|/ |"|     |" ___|    ___     | \ |"|  |  _"\U |  _"\ u  
   |_"_|   <|  \| |>/| | | | |  _|" U | | u  U| |_  u   |_"_|   <|  \| |>/| | | |\| |_) |/  
    | |    U| |\  |uU| |_| |\| |___  \| |/__ \|  _|/     | |    U| |\  |uU| |_| |\|  _ <    
  U/| |\u   |_| \_|  |____/ u|_____|  |_____| |_|      U/| |\u   |_| \_|  |____/ u|_| \_\   
.-,_|___|_,-.||   \\,-.|||_   <<   >>  //  \\  )(\\,-.-,_|___|_,-.||   \\,-.|||_   //   \\_  
 \_)-' '-(_/ (_")  (_/(__)_) (__) (__)(_")("_)(__)(_/ \_)-' '-(_/ (_")  (_/(__)_) (__)  (__)
                                                             
}"
                                                              
cat("\n \n \n#############################################################################################\n")

cat(logo)

cat("
#############################################################################################\n
\n
Thank you for using INDELfindR:\nA tool for detecting somatic insertion deletions\nwritten in R v4.1.0\n
\n

#############################################################################################\n \n")

#############################################################################################
#
# Define Args:
#
#############################################################################################

suppressMessages(library(argparse))

# create parser object
parser <- ArgumentParser()

# specify our desired options 
# by default ArgumentParser will add an help option 
# turn off for dev:

parser$add_argument("-v", "--verbose", action="store_true", default=F,
                    help="Print extra output [default]")
parser$add_argument("-q", "--quietly", action="store_false", 
                    dest="verbose", help="Print little output")
parser$add_argument("-a", "--alignment_file", type="character",required=T, 
                    help="Filepath to alignment file in bam format",
                    metavar="filename")
parser$add_argument("-f", "--flanking_region_length", type="integer", default=10, 
                    help="Minimum number of `=` or `X` CIGAR operators required to flank a string of nearby `I` `D` operators in order for a call to be made (default = 10).",
                    metavar="number")
parser$add_argument("-b", "--bam_bin_size", type="integer", default=100000, 
                    help="Length of non-overlapping sliding windows to use to extract overlapping reads from bam file and store in memory at a time (default = 100000). Larger windows require loading more reads into memory at one time, while smaller windows require longer runtime due to evaluating more reads which overlap neighboring sliding windows and removing more duplicate read calls",
                    metavar="number")
parser$add_argument("-l", "--min_indel_length", type="integer", default=3, 
                    help="Minumum number of ref or alt allele basepairs for an indel to be called (default = 3)",
                    metavar="number")
parser$add_argument("-nr", "--number_reads", type="integer", default=4, 
                    help="Minumum number of supporting reads for a single given indel to be called (default = 4)",
                    metavar="number")
parser$add_argument("-mq", "--mapq_filter", type="integer", default=20, 
                    help="Minumum read mapping quality required to evaluate a read (default = 20).",
                    metavar="number")
parser$add_argument("-t", "--target_regions", type="character", default=F, 
                    help="File path to .bed file containing regions in which to perform variant calling. Chromosome syntax must match bed file (ex. `Chr1`, `chr1`, or `1`).",
                    metavar="filename")
parser$add_argument("-nc", "--number_cores", type="integer", default=4, 
                    help="Number of cores to use for fork/join parallel computation during indel calling and filtering (default = 4).",
                    metavar="number")
parser$add_argument("-p", "--primary_chromosomes", action="store_true", default=T, 
                    help="Call indels in primary chromosomes only ignoring ALT contig alignments (chr1-22,X,Y,M only) (default = True)."
)
parser$add_argument("-vaf", "--vaf_filter", type="double", default=0.01, 
                    help="Minumum variant allele freqency required for an indel to be reported (default = 0.01).",
                    metavar="number")
parser$add_argument("-dp", "--read_depth_filter", type="double", default=10, 
                    help="Minumum indel range read depth required for an indel to be reported (default = 20).",
                    metavar="number")
# parser$add_argument("-z", "--zero_based", action="store_true",
#                     help="convert variant calls from zero based to one based")
parser$add_argument("-o", "--outname", default=getwd(),
                    help="Define the output directory path",metavar="filepath")

args <- parser$parse_args()

bamPath <- args$alignment_file
bam_region_bin_size <- args$bam_bin_size
verbose_arg <- args$verbose
flanking_region_length <- args$flanking_region_length
target_regions <- args$target_regions
number_cores <- args$number_cores
primary_chromosomes <- args$primary_chromosomes
min_indel_length <- args$min_indel_length
mapq_threshold <- args$mapq_filter
min_supporting_reads <- args$number_reads
min_vaf <- args$vaf_filter
min_read_depth <- args$read_depth_filter
#zero_based <- args$zero_based
outname <- args$outname

# #To run during dev:
# bamPath <- "/Users/George/indel_detection_tool_project/bam_files_for_testing/EGFR_mutations_reference_dwgsim.sorted.bam"
# #bamPath <- "/Users/George/indel_detection_tool_project/benchmarking/indelfindr_results/miss_validate_igv/miss_5.sorted.bam"
# bam_region_bin_size <- 100000000 #dev-on
# verbose_arg <- FALSE
# flanking_region_length <- 10
# #target_regions <- F
# target_regions <- "/Users/George/indel_detection_tool_project/data_for_testing/EGFR_regions_of_interest.txt"
# #target_regions <- "/Users/George/indel_detection_tool_project/indelfindr/dev_scripts/complex_indel_bam.bed"
# number_cores <- 1 # Make default 2
# primary_chromosomes <- T
# min_indel_length <- 3
# mapq_threshold <- 20
# min_supporting_reads <- 4
# min_vaf <- 0.01
# min_read_depth <- 10
# #zero_based <- F
# outname <- "/Users/George/indel_detection_tool_project/test_output_dir/test_vcf_output"

# #bamPath <- "/Users/George/indel_detection_tool_project/bam_files_for_testing/ebi_09-09_chr2_subset.bam"
# #bamPath <- "/Users/George/indel_detection_tool_project/bam_files_for_testing/bam_for_testing_1656854-1657354.bam"
# bamPath <- "/Users/George/indel_detection_tool_project/bam_files_for_testing/EGFR_mutations_reference_dwgsim.sorted.bam"
# #bamPath <- "/Users/George/indel_detection_tool_project/bam_files_for_testing/test_read.bam"
# #max_window_length <- 5 #dev-on
# bam_region_bin_size <- 1000000 #dev-on
# verbose_arg=FALSE
# #window_load_length <- 500
# flanking_region_length <- 10
# #sliding_window_size <- 20
# #Optional parameters for testing:
# target_regions <- "/Users/George/indel_detection_tool_project/data_for_testing/EGFR_regions_of_interest.txt"
# #target_regions <- F
# #target_regions <- "/Users/George/indel_detection_tool_project/data_for_testing/debug_10_27.txt"
# number_cores <- 2 # Make default 2
# primary_chromosomes <- T
# min_indel_length <- 3
# mapq_threshold <- 20
# min_supporting_reads <- 3
# min_vaf <- 0.01
# min_read_depth <- 10

#############################################################################################
#
# Load Packages
#
#############################################################################################

suppressMessages(library(GenomicAlignments))
suppressMessages(library(plyr))
suppressMessages(library(tidyverse))
suppressMessages(library(GenomicFeatures))
suppressMessages(library(BSgenome.Hsapiens.UCSC.hg38)) 
#suppressMessages(library(bamsignals)) 
suppressMessages(library(inline))
suppressMessages(library(bettermc))
suppressMessages(library(stringr))

#############################################################################################
#
# Load Functions:
#
#############################################################################################

# Replace the path below with the path to your /indelfindr/src/functions.R file.
source("/Users/George/indel_detection_tool_project/indelfindr/src/functions_bug_fix_march_28.R")
#source("/gpfs/data/dgamsiz/Uzun_Lab/gtollefs/indel_detection_project/running_directory/functions.R")

#############################################################################################
#
# Load Utility Data
#
#############################################################################################

options(scipen=20)

# get all hg38 chromosome lengths
chg38_chromosome_lengths <- getChromInfoFromUCSC("hg38") # all hg38 chr/altcontig lengths  (595 chromosomes/contigs in hg38 annotation)

# load hg38 reference genome sequence data (688MB) - add option to load this optionally, if user opts to use BBtools Reformat tool, or other aligner, for extended cigar string, it will cut down on runtime and not require loading reference genome sequence into memory but this would require we dev a handler for refined cigar bam input.
hg38_genome <- BSgenome.Hsapiens.UCSC.hg38

#############################################################################################
#
# Running Script:
#
#############################################################################################

# define data table for reporting all refined cigar string frames in sliding windows
master_indel_record_table <- c("chr","start_pos",
                               "end_pos",
                               "refined_cigar_string",
                               "collapsed_cigar_string",
                               "reference_allele",
                               "alt_allele",
                               "strand",
                               "read_name")  %>% purrr::map_dfc(setNames, object = list(as.character()))

# Subset primary chromosomes only for extracting reads for indel calling (optional)
if (primary_chromosomes == T){
  #chr_in_bam <- get_primary_chroms(chr_in_bam)
  chr_in_bam <-  c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9",'chr10',"chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22","chrX","chrY","chrM")
}

if (target_regions != FALSE) {
  # load target regions
  target_regions_table <- read.table(target_regions,sep="\t",col.names = c("chr","start","stop"))
  chr_in_bam <- unique(target_regions_table$chr)
}

for (each_chromosome in chr_in_bam){

  message(paste("Analyzing chromosome:",each_chromosome))
  
  # Define chr subset reference sequence
  hg38_genome_chr_subset <- hg38_genome[[each_chromosome]] 

    if (target_regions != FALSE){
    
    for (regions_table_row in 1:nrow(subset(target_regions_table,chr=chr_in_bam))){
      
      min_pos <- subset(target_regions_table,chr=chr_in_bam)$start[regions_table_row]
      max_pos <- subset(target_regions_table,chr=chr_in_bam)$stop[regions_table_row]
      
      target_bin_size <- bam_region_bin_size
      
      if (max_pos-min_pos < bam_region_bin_size){
        target_bin_size <- max_pos-min_pos
      }
      
      intervals <- seq_with_uneven_last(from=min_pos,to=max_pos,by=target_bin_size)
      
      # define coordinates of sliding windows
      sliding_windows_per_bam_region=data.frame(chr=each_chromosome,
                                                start=intervals[-(length(intervals))],
                                                end=intervals[-1])
      
      per_bam_region_indel_records <- c("chr",
                                        "start_pos",
                                        "end_pos",
                                        "refined_cigar_string",
                                        "collapsed_cigar_string",
                                        "reference_allele",
                                        "alt_allele",
                                        "strand",
                                        "read_name")  %>% purrr::map_dfc(setNames, object = list(as.character()))
      
      # Find indels in each bam region using all available cores in parallel
      per_bam_region_indel_records <-
          do.call(
            rbind, bettermc::mclapply(1:nrow(sliding_windows_per_bam_region),run_algo_all_reads_each_bam_region_with_simple_indel_padding,per_bam_region_indel_records,mc.cores=number_cores,mc.preschedule = F))
      
      per_chrom_filtered_calls <- filter_and_annotate_calls(per_bam_region_indel_records)
      
      ## Add each chromosome results to the master results table
      master_indel_record_table <- rbind(master_indel_record_table,per_chrom_filtered_calls)
    
      wait()
      
     } # end each targeted region in bed file iteration
    
  } else if (target_regions == FALSE){ #end targeted region mode conditional 
    
    # do all analysis per chromosome, per bam region chunk alignment region, per read in alignment region
    
    size_chr <- chg38_chromosome_lengths[which(chg38_chromosome_lengths[,1]==each_chromosome),2]
    
    # p_for_chr_extract <- ScanBamParam(which=GRanges(
    #   Rle(each_chromosome),
    #   IRanges(1, size_chr)),
    #   what=c("pos"))
    
    # max_pos<- max(na.omit(unlist(scanBam(bamPath, param=p_for_chr_extract)))) # add na omit for finding max read start when there are NA positions returned - find out why and decide if should remove them. 
    # min_pos<- min(na.omit(unlist(scanBam(bamPath, param=p_for_chr_extract)))) # add na omit for finding max read start when there are NA positions returned - find out why and decide if should remove them. 
    
    # get total number reads 
    #overlap.counts <- suppressMessages((bamCount(bamPath,GRanges(Rle(each_chromosome,1), IRanges(start=min_pos, end=max_pos)))))
    
    target_bin_size <- bam_region_bin_size
    
    # if (max_pos-min_pos < bam_region_bin_size){
    #   target_bin_size <- max_pos-min_pos
    # }
    
    intervals <- seq_with_uneven_last(from=1,to=size_chr,by=target_bin_size)
    
    # define coordinates of sliding windows
    sliding_windows_per_bam_region=data.frame(chr=each_chromosome,
                                              start=intervals[-(length(intervals))],
                                              end=intervals[-1])
    
    #rm(p_for_chr_extract)
    
    # Initialize dataframe for collecting each bam region indel calls
    per_bam_region_indel_records <- c("chr",
                                      "start_pos",
                                      "end_pos",
                                      "refined_cigar_string",
                                      "collapsed_cigar_string",
                                      "reference_allele",
                                      "alt_allele",
                                      "strand",
                                      "read_name")  %>% purrr::map_dfc(setNames, object = list(as.character()))
    
    # Find indels in each bam region using all available cores in parallel

    per_bam_region_indel_records <-
      do.call(
        rbind, bettermc::mclapply(1:nrow(sliding_windows_per_bam_region),run_algo_all_reads_each_bam_region_with_simple_indel_padding,per_bam_region_indel_records,mc.cores=number_cores,mc.preschedule = F))

    per_chrom_filtered_calls <- filter_and_annotate_calls(per_bam_region_indel_records)
    
    master_indel_record_table <- rbind(master_indel_record_table,per_chrom_filtered_calls)
    
    wait()
    
  } # end genome wide mode conditional
  
} # end chr iteration

message(paste0("Saving ",nrow(master_indel_record_table), " indels to output filepath: ",outname,"."))

# # make directory
# suppressWarnings(
#   "suppressWarnings("In dir.create(file.path("/Users/George/indel_detection_tool_project/test_output_dir/")) :
#   '/Users/George/indel_detection_tool_project/test_output_dir' already exists")"
# )

#dir.create(file.path(outname)) # suppress warnings

# write out VCF file
write.vcf(master_indel_record_table,phased=FALSE,paste0(outname,".vcf"))

#write out data table with cigar string
write.table(master_indel_record_table,paste0(outname,".indel.cigars.csv"),sep=",",row.names=F)

message("Run Complete")
